test/runtests.jl

```diff
@@ -1128,19 +1151,34 @@ Base.similar(bc::Broadcast.Broadcasted{Broadcast.ArrayStyle{MyArray}}, ::Type{El
     # used inside of broadcast but we also test it here explicitly
     @test isa(@inferred(Base.dataids(s)), NTuple{N, UInt} where {N})
 
-    s = StructArray{ComplexF64}((MyArray(rand(2)), MyArray(rand(2))))
-    @test_throws MethodError s .+ s
+    # Make sure we can handle style with similar defined
+    # And we can handle most conflicts
+    # `s1` and `s2` have similar defined, but `s3` does not
+    # `s2` conflicts with `s1` and `s3` and is weaker than `DefaultArrayStyle`
+    s1 = StructArray{ComplexF64}((MyArray1(rand(2)), MyArray1(rand(2))))
+    s2 = StructArray{ComplexF64}((MyArray2(rand(2)), MyArray2(rand(2))))
+    s3 = StructArray{ComplexF64}((MyArray3(rand(2)), MyArray3(rand(2))))
+    s4 = StructArray{ComplexF64}((rand(2), rand(2)))
+
+    function _test_similar(a, b, c)
+        try
+            d = StructArray{ComplexF64}((a.re .+ b.re .- c.re, a.im .+ b.im .- c.im))
+            @test typeof(a .+ b .- c) == typeof(d)
+        catch
+            @test_throws MethodError a .+ b .- c
+        end
```

Comment:
This had escaped me before, but I'm wondering: could it be possible to be explicit here on which it is (correct result or method error) based on the input types?

Ideally one would want to explicitly test what one is getting, so I would suggest to remove the helper function `_test_similar` and just write something like

```julia
if s2 in (s, s′, s″) && (s1 in (s, s′, s″) || s3 in (s, s′, s″))
    # test method error
else
    # test correct result
end
```

in the loop body.

(I'm not sure whether that's the correct criterion.)